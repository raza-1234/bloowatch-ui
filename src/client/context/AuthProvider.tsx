import { createContext, useState, useEffect, useContext } from "react";
import api from "../axios/api";
import { AxiosResponse } from "axios";
import React from "react";
import { jwtDecode } from "jwt-decode";
import { Decoded_Token, UserDetail, Children, AuthContexType } from "../types/types";

const AuthContext = createContext<AuthContexType | undefined>(undefined);

export const AuthProvider = ({children}: Children) => {

  const [userData, setUserData] = useState<UserDetail>({
    accessToken: localStorage.getItem("access_token") || null,
    email: "",
    id: undefined,
    name: ""
  })

  useEffect(() => {      
    if (userData?.accessToken){
      const data: any = jwtDecode(userData.accessToken);
      if ( data.userId ) {        
        getUserData(userData.accessToken, data.userId);
      }
    }
  }, [userData?.accessToken])

  const getUserData = async (accessToken: string | null, userId?: number): Promise<UserDetail|void> => {
    try {
      const response: AxiosResponse = await api.get(`user-detail/${userId}`,
        { 
          headers: {
            "Authorization" : `Bearer ${accessToken}`
          }
        }
      )      
      setUserData({...userData, ...response.data});
    } catch (err){
      console.log(err);
    }
  }

  return(
    <AuthContext.Provider value={{ getUserData, userData, setUserData}}>
      {children}
    </AuthContext.Provider>
  )
}

const useAuthData = () => {
  const context = useContext(AuthContext)
  if(!context) {
    throw new Error('AuthData must in the AuthProvider')
  }

  return context;
}

export default useAuthData