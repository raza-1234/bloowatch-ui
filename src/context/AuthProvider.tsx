import { createContext, useState, useEffect, useContext } from "react";
import api from "../axios/api";
import { AxiosResponse } from "axios";
import React from "react";
import { jwtDecode } from "jwt-decode";
import { Decoded_Token, UserDetail, Children, AuthContexType } from "../types/types";

const AuthContext = createContext<AuthContexType | undefined>(undefined);

export const AuthProvider = ({children}: Children) => {

  const [userData, setUserData] = useState<UserDetail>({
    accessToken: localStorage.getItem("access_token") || null,
    email: "",
    id: undefined,
    name: ""
  })

  useEffect(() => {  
    if (userData?.accessToken){
      const { userId }: Decoded_Token = jwtDecode(userData.accessToken);
      if ( userId ) {
        getUserData(userData.accessToken, userId);
      }
    }
  }, [userData?.accessToken])

  const getUserData = async (accessToken: string, userId: number): Promise<UserDetail|void> => {
    try {
      const response: AxiosResponse = await api.get(`http://localhost:3500/user-detail/${userId}`,
        { 
          headers: {
            "Authorization" : `Bearer ${accessToken}`
          }
        }
      )      
      setUserData({...userData, ...response.data});
    } catch (err){
      console.log(err);
    }
  }

  return(
    <AuthContext.Provider value={{ getUserData, userData, setUserData}}>
      {children}
    </AuthContext.Provider>
  )
}

const AuthData = () => {
  const context = useContext(AuthContext)
  console.log('context', context)
  if(!context) {
    throw new Error('AuthData must in the AuthProvider')
  }

  return context;
}

export default AuthData