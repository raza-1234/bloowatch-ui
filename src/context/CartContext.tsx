import React, {createContext, useContext, useEffect, useState} from "react";
import { AxiosResponse } from 'axios';
import { Children, CartContextType, CartContextProvider, STATUS_TEXT } from "../types/types";
import api from '../axios/api';
import { jwtDecode } from "jwt-decode";
import AuthData from "./AuthProvider";

const CartContext = createContext<CartContextProvider | null>(null);

export const CartProvider = ({children}: Children) => {
  const [cart, setCart] = useState<CartContextType>({
    cartData: [],
    cartCount: 0,
  })
  
  const { userData } = AuthData();

  useEffect(() => {   
    if (userData?.accessToken){
      const { userId }: {userId: number} = jwtDecode(userData?.accessToken);
      if ( userId ) {
        fetchCartProducts(userData?.accessToken, userId)
      }
    } 
  },[userData?.accessToken])

  const fetchCartProducts = async (accessToken?: string, userId?: number): Promise<void> => {
    try {
      const response: AxiosResponse = await api.get(`cart/getAllCartProducts/${userId}`, 
        {
          headers: {"Authorization" : `Bearer ${accessToken}`
        }
      })      
      if (response.statusText === STATUS_TEXT && response.data.length){
        setCart({cartData: response.data, cartCount: response.data.length});
        return;
      }
      setCart({cartData: [], cartCount: 0});
    } catch (err: any){
      console.log(err);
    }
  }

  return (
    <CartContext.Provider value = {{cart, fetchCartProducts}} >
      {children}
    </CartContext.Provider>
  )
}

const CartContextData = () => {
  const context = useContext(CartContext);
  if(!context) {
    throw new Error('Cart Context Not Setup Properly.')
  }

  return context;
}
export default CartContextData;


