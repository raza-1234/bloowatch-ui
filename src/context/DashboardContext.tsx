import React, { FC, createContext, useState} from 'react'
import { Product, Paging, PRODUCT_STATUS, RESPONSE_TEXT, DashboardContextValue, Children } from '../types/types';
import { AxiosResponse } from 'axios';
import api from '../axios/api';
import useAuth from '../hooks/useAuth';
import logOut from '../utils/logout';

export const dashboardContext = createContext<DashboardContextValue | null>(null);
const DashboardContext = ({children}: Children) => {

  const [ products, setProducts ] = useState<Product[]>([])
  const [pagingInfo, setPagingInfo] = useState<Paging>()
  const [page, setPage] = useState<number>();
  const [category, setCategory] = useState("")
  const [search, setSearch] = useState("")
  const [price, setPrice] = useState<number[]>([0, 100]);
  const { auth, setAuth }: any = useAuth();  

  async function fetchProducts(productPrice: number[], productCategory?: string, title?: string, pageNumber?: number): Promise<void> {       
    try {
      let response: AxiosResponse;
      if (title?.trim()){
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}&search=${title.toLowerCase()}`, {headers: {"Authorization" : `Bearer ${auth.token}`}}
        );
      }
      else if (productCategory?.trim()){
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}&category=${productCategory.toLowerCase()}`, {headers: {"Authorization" : `Bearer ${auth.token}`}});
      }
      else {
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}`, {headers: {"Authorization" : `Bearer ${auth.token}`}});
      }

      if (response.data.message === PRODUCT_STATUS){
        setProducts([])
        setPagingInfo(undefined)
        return ; 
      }            
      setProducts(response.data.data)
      setPagingInfo(response.data.paging)
    } catch (err){
      console.log(err);
      if (err.response.data === RESPONSE_TEXT){
        logOut();
        setAuth();
      }
    }
  }  

  return (
    <dashboardContext.Provider 
      value={
        {products, pagingInfo, fetchProducts, page, setPage, category, setCategory, search, setSearch, setPrice, price  }
      }
    >
      {children}
    </dashboardContext.Provider>
  )
}

export default DashboardContext
