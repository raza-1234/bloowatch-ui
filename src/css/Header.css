
.bloowatch-header__wrapper {
  width: 100%;
  margin: 0 auto;
  padding-top: 0.5%;
  padding-bottom: 0.5%;
  box-shadow: 0px 1px 6px black;
}

.bloowatch-header__wrapper img {
  width: 150px;
}

.bloowatch-header__content {
  display: flex;
  justify-content: space-between;
  margin: 1.5%;
}

.bloowatch-header__nav-list {
  list-style-type: none;
  display: flex;
  justify-content: space-between;
}

.bloowatch-header__nav-list-item {
  margin-right: 30px;
  color: black;
  text-decoration: none;
}

.bloowatch-header__user {
  position: relative;
  color: rgb(2, 29, 175);
}

.bloowatch-header__user-drop-down {
  border: 2px solid rgb(2, 29, 175);
  border-radius: 5px;
  display: flex;
  flex-direction: column;
  width: 180px;
  position: absolute;
  right: 50%;
  top: 100%;
  background-color: rgba(2, 29, 175, 0.4);
}

.bloowatch-header__user-drop-down hr{
  width: 100%;
  border: 1px solid white;
}

.bloowatch-header__user-drop-down div {
  padding: 8px;
  text-align: left;
  font-weight: 600;
}

.bloowatch-header__user h4{
  cursor: pointer;
  text-transform: capitalize;
}
