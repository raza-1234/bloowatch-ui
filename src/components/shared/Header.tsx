import React from 'react'
import { useState } from 'react'
import logo from "../../assets/logo.png"
import { Link } from 'react-router-dom'
import "../../css/Header.css"
import useAuth from '../../hooks/useAuth'
import logOut from '../../utils/logout'

const Header = () => {
  const { auth, setAuth }:any = useAuth();
  const [ showDropDown, setShowDropDown ] = useState(false);

  const LogOut = (): void => {
    logOut();
    setAuth();
    setShowDropDown(!showDropDown)
  }

  return (
    <header className='bloowatch-header__wrapper'>
      <div className='bloowatch-header__content'>
        <img src={logo} alt='logo'/>
        <div className='bloowatch-header__nav-list'>
          <li>
            <Link className = "bloowatch-header__nav-list-item" to = "/shop">Shop</Link>
          </li>          
          <li>
            <Link className = "bloowatch-header__nav-list-item" to = "/cart">Cart</Link>
          </li>
          { auth?.userEmail ? 
            <div className='bloowatch-header__user' onClick={() => setShowDropDown(!showDropDown)}>
              <li> 
                <h4>{auth?.userName}</h4> 
              </li>
              {
                showDropDown && 
                <div className='bloowatch-header__user-drop-down'>
                  <div className='bloowatch-header__edit-profile'>
                    <Link className = "bloowatch-header__nav-list-item" to = "/edit-user">Edit Profile</Link>
                  </div>
                  <hr/>
                  <div className='bloowatch-header__logout'>
                    <Link className = "bloowatch-header__nav-list-item" to = "/" onClick={LogOut}>Log Out</Link>
                  </div>
                </div>
              }
            </div>

            :
            <React.Fragment>
              <li>
                <Link className = "bloowatch-header__nav-list-item" to = "/register">Register</Link>
              </li>
              <li>
                <Link className = "bloowatch-header__nav-list-item" to = "/">Login</Link>
              </li>
            </React.Fragment>
          }
        </div>
      </div>
    </header>
  )
}

export default Header
