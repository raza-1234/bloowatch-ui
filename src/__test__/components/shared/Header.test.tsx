import React from 'react';
import { render, screen, fireEvent, renderHook } from '@testing-library/react';
import Header from '../../../components/shared/Header';
import { BrowserRouter } from 'react-router-dom';
import AuthData, {AuthProvider} from '../../../context/AuthProvider';
import {CartProvider} from '../../../context/CartContext';

// jest.mock('../../../context/AuthProvider', () => ({
  // ...jest.requireActual("../../../context/AuthProvider"),

  // default: ({AuthData: jest.fn()})
  // AuthData: jest.fn().mockReturnValue({
  //   userData: {},
  //   setUserData: jest.fn(),
  //   getUserData: jest.fn()
  // })
// }));

  // AuthData: jest.fn().mockReturnValue({
  //   userData: {},
  //   setUserData: jest.fn(),
  //   getUserData: jest.fn()
  // })



// jest.mock('../../../context/CartContext', () => ({
//   ...jest.requireActual("../../../context/CartContext"),
//   CartContextData: jest.fn().mockReturnValue({ 
//     cart: {
//       cartData: [],
//       count: 0
//     },
//     fetchCartProducts: jest.fn()
//   })
// }));


// jest.mock('../../../context/AuthProvider', () => ({
//   __esModule: true, 
//   default: jest.fn(() => ({
//     userData: {
//       accessToken: 'mockAccessToken',
//       name: 'Mock User',
//     },
//     setUserData: jest.fn(),
//   })),
// }));

// jest.mock('../../../context/CartContext', () => ({
//   __esModule: true,
//   default: jest.fn(() => ({
//     cart: {
//       cartData: [],
//       cartCount: 0,
//     },
//   })),
// }));

const mockUserData = {
  accessToken: "fakeToken",
  email: "fakeEmail@gmail.com",
  id: 1,
  name: "fake name",
}

const buildApp = () => {
  return (
    render(
      <BrowserRouter>
        <AuthProvider>
          <CartProvider>
            <Header/>  
          </CartProvider>
        </AuthProvider>
      </BrowserRouter>
    )
  )
}

describe("Header", () => {

  it("should render successfully", () => {
    // mockAuthData.mockReturnValueOnce({getUserData: jest.fn(), userData: {
    //   accessToken: "fakeToken",
    //   email: "fakeEmail@gmail.com",
    //   id: 1,
    //   name: "fake name",
    // }, setUserData: jest.fn() })

    const {container} = buildApp();

    expect(container).toMatchSnapshot(); 
    // const { result, waitForTheNextUpdate } = renderHook(() => AuthData());
    // console.log('result', result.current.getUserData({
    //   accessToken: "fakeToken",
    //   email: "fakeEmail@gmail.com",
    //   id: 1,
    //   name: "fake name",
    ;

    // act(() => {
    //   mount<Header>
    // })
 
    // const header_wrapper = screen.getByTestId("header_wrapper");
    // const header_content = screen.getByTestId("header_content")

    // expect(header_wrapper).toBeInTheDocument();
    // expect(header_wrapper).toHaveClass("bloowatch-header__wrapper");
    // expect(header_content).toBeInTheDocument();
    // expect(header_content).toHaveClass("bloowatch-header__content");
  }) 
 
  // it("should render bloowatch logo", () => {
  //   buildApp();

  //   const logo = screen.getByTestId("logo");
  //   const images = screen.getAllByRole("img")

  //   expect(logo).toBeInTheDocument();
  //   expect(logo).toHaveClass("bloowatch-header__logo");
  //   expect(images[0]).toHaveProperty("alt", "logo");
  //   expect(images[0]).toHaveProperty("src", "http://localhost/logo.png");
  //   fireEvent.click(images[0]);
  //   expect(window.location.pathname).toBe("/shop");
  // })

  // it("should render headers menu", () => {
  //   buildApp();

  //   const header_menu = screen.getByTestId("header_menu");
  //   const hamburger_menu = screen.getByTestId("hamburger_menu");
  //   const hamburger_menu_icon = screen.getByTestId("hamburger_menu_icon");
  //   const shopPage = screen.getByText("Shop");
  //   const registerPage = screen.getByText("Register");
  //   const loginPage = screen.getByText("Login");

  //   expect(header_menu).toBeInTheDocument();
  //   expect(hamburger_menu).toBeInTheDocument();
  //   expect(hamburger_menu_icon).toBeInTheDocument();
  //   expect(shopPage).toBeInTheDocument();
  //   expect(registerPage).toBeInTheDocument();
  //   expect(loginPage).toBeInTheDocument();

  //   expect(header_menu).toHaveClass("bloowatch-header__nav-list");
  //   expect(hamburger_menu).toHaveClass("bloowatch-header__mobile-view");

  //   fireEvent.click(shopPage);
  //   expect(window.location.pathname).toBe("/shop");

  //   fireEvent.click(registerPage);
  //   expect(window.location.pathname).toBe("/register");

  //   fireEvent.click(loginPage);
  //   expect(window.location.pathname).toBe("/");

  //   fireEvent.click(hamburger_menu_icon);
  //   const mobile_screen_wrapper = screen.getByTestId("mobile_screen_wrapper");
  //   expect(header_menu).toHaveClass("bloowatch-header__mobile-menu");
  //   expect(mobile_screen_wrapper).toBeInTheDocument();
  //   expect(mobile_screen_wrapper).toHaveClass("bloowatch-mobile-screen__wrapper");
  // })

  // it("should do whatever i say", () => {
  //   const { container } = buildApp();
    
  //   expect(container).toMatchSnapshot();
  //   expect(AuthData).toHaveBeenCalled();
  // })
 
})