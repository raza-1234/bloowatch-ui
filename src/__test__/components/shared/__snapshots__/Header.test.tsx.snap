// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render successfully 1`] = `
<div>
  <header
    class="bloowatch-header__wrapper"
    data-testid="header_wrapper"
  >
    <div
      class="bloowatch-header__content"
      data-testid="header_content"
    >
      <div
        class="bloowatch-header__logo"
        data-testid="logo"
      >
        <a
          href="/shop"
        >
          <img
            alt="logo"
            src="logo.png"
          />
        </a>
      </div>
      <div
        class="bloowatch-header__nav-list"
        data-testid="header_menu"
      >
        <li>
          <a
            class="bloowatch-header__nav-list-item"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li>
          <a
            class="bloowatch-header__nav-list-item"
            href="/register"
          >
            Register
          </a>
        </li>
        <li>
          <a
            class="bloowatch-header__nav-list-item"
            href="/"
          >
            Login
          </a>
        </li>
      </div>
      <div
        class="bloowatch-header__mobile-view"
        data-testid="hamburger_menu"
      >
        <svg
          data-testid="hamburger_menu_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 96v64h448V96H32zm0 128v64h448v-64H32zm0 128v64h448v-64H32z"
          />
        </svg>
      </div>
    </div>
  </header>
</div>
`;
